src/replication.c

```diff
@@ -3216,14 +3222,44 @@ void setupMainConnForPsync(connection *conn) {
                                      "accept connections in read-write mode.\n");
         }
         dualChannelSyncHandlePsync();
-        return;
+        return C_OK;
     }
 
-error:
+    return psync_result;
+}
+
+/* Replication: Replica side.
+ * This connection handler fires after rdb-connection was initialized. We use it
+ * to adjust the replica main for loading incremental changes into the local buffer. */
+void setupMainConnForPsync(connection *conn) {
+    char *err = NULL;
+    int ret;
+
+    switch (server.repl_state) {
+    case REPL_STATE_SEND_HANDSHAKE: ret = dualChannelReplicationMainConnSendHandshake(conn, &err); break;
+    case REPL_STATE_RECEIVE_CAPA_REPLY:
+        ret = dualChannelReplicationMainConnReceiveCapaReply(conn, &err);
+        if (ret == C_ERR) {
+            break;
+        }
+        sdsfree(err);
+        err = NULL;
+        /* fall through */
+    case REPL_STATE_SEND_PSYNC: ret = dualChannelReplicationMainConnSendPsync(conn, &err); break;
+    case REPL_STATE_RECEIVE_PSYNC_REPLY: ret = dualChannelReplicationMainConnReceivePsyncReply(conn, &err); break;
+    default:
+        serverLog(LL_WARNING, "Unexpected replication state: %d", server.repl_state);
+        ret = C_ERR;
+        break;
+    }
     sdsfree(err);
+    if (ret == C_ERR) goto error;
+    return;
+
+error:
     /* The dual-channel sync session must be aborted for any psync_result other than PSYNC_CONTINUE or PSYNC_WAIT_REPLY.
      */
-    serverLog(LL_WARNING, "Aborting dual channel sync. Main channel psync result %d", psync_result);
+    serverLog(LL_WARNING, "Aborting dual channel sync. Main channel psync result %d", ret);
     cancelReplicationHandshake(1);
```

Comment:
`goto` is not really necessary here. With the refactoring, we could just run the function without branching, which is even better.

```suggestion
    if (ret != C_OK) {
        serverLog(LL_WARNING, "Aborting dual channel sync. Main channel psync result %d", ret);
        cancelReplicationHandshake(1);
    }
```